erc-3156 ready for review (#3181)
ethereum/EIPs: EIPS/eip-3156.md

### Lender Specification

A `lender` smart contract implementing a flash lending feature MUST implement a `maxFlashAmount` function:
```

> function maxFlashAmount(address token) external external returns (uint256);

Reviewer: ```suggestion
function maxFlashAmount(address token) external returns (uint256);
```

Reviewer: A couple others like this

### Lender Specification

A `lender` MUST implement the IERC3156FlashLender interface.
```
interface IERC3156FlashLender {
    function maxFlashAmount(address token) external view returns (uint256);
    function flashFee(address token, uint256 amount) external view returns (uint256);
    function flashLoan(
        address receiver,
        address token,
        uint256 amount,
        bytes calldata data
    ) external;

> }

Author: I've seen often as a style to define functions in a single line when the number of arguments is three or fewer, and expand them to an argument per line with four or more.

I've gone for each argument in a line in function definitions.

The `lender` smart contract MUST include `msg.sender` as the first argument to `onFlashLoan`.

The `lender` smart contract MUST NOT modify the `data` parameter received to `flashLoan`, and must include it as the last argument to `onFlashLoan`.

`amount + fee` tokens MUST be returned to the `lender` before the end of the `flashLoan` function, reverting otherwise.

The return of the loan MUST be initiated by the `receiver`. The `lender` MUST NOT take the `amount + fee` from the `receiver` or any other contract.

### Receiver Specification

A `receiver` of flash loans MUST implement an `onFlashLoan` callback:

```
interface FlashBorrowerLike {
    function onFlashLoan(address sender, address token, uint256 amount, uint256 fee, bytes calldata) external;
}
```

For the transaction to not revert, `receiver` MUST send `amount + fee` of `token` to `msg.sender` before the end of `onFlashLoan`.

## Rationale

The interfaces described in this ERC have been chosen as to cover the known flash lending use cases, while allowing for safe and gas efficient implementations.

`flashFee(address token, uint256 amount)`

> `flashFee` reverts on unsupported contracts, because returning a numerical value would be incorrect.

Reviewer: ```suggestion
`flashFee` reverts on unsupported tokens, because returning a numerical value would be incorrect.
```

}
```

The `flashLoan` function MUST transfer `amount` `token` to `receiver` before calling `FlashBorrowerLike(receiver).onFlashLoan(...)`.

The `lender` smart contract MUST include `msg.sender` as the first argument to `onFlashLoan`.

The `lender` smart contract MUST NOT modify the `data` parameter received to `flashLoan`, and must include it as the last argument to `onFlashLoan`.

The `lender` MUST take the `amount + fee` `token` from the `receiver`, or revert if this is not succesful.

### Receiver Specification

A `receiver` of flash loans MUST implement an `onFlashLoan` callback:

```

> interface FlashBorrowerLike {

Author: ```suggestion
interface IERC3156FlashBorrower {
```

```
function batchFlashLoan(address receiver, address[] calldata token, uint256[] calldata amount, bytes calldata data) external {
  ...
  IERC3156BatchFlashBorrower(receiver).onBatchFlashLoan(msg.sender, tokens, amounts, fees, data);
  ...
}
```

For each `token` in `tokens`, the `batchFlashLoan` function MUST transfer `amounts[i]` of `tokens[i]` to `receiver` before the callback to the borrower.

The `batchFlashLoan` function MUST include `msg.sender` as the `sender` to `onBatchFlashLoan`.

The `batchFlashLoan` function MUST include a `fees` argument to `onBatchFlashLoan` with the fee to pay for each individual `token` and `amount` lent.

The `batchFlashLoan` function MUST NOT modify the `tokens`, `amounts` and `data` parameters received, and MUST pass them on to `onBatchFlashLoan`.

After the callback, for each `token` in `tokens`, the `batchFlashLoan` function MUST take the `amounts[i] + fees[i]` of `tokens[i]` from the `receiver`, or revert if this is not succesful.

### Receiver Specification

A `receiver` of flash loans MUST implement an `onFlashLoan` callback:

```
interface IERC3156FlashBorrower {
    function onFlashLoan(

>         address sender,

Author: I also refactored from `address` to `IERC20` wherever possible, did I go too far?

The `amount` will be required in the `onFlashLoan` function, which the lender took as a parameter. An alternative implementation which would embed the `amount` in the `data` parameter by the caller would require an additional mechanism for the receiver to verify its accuracy, and is not advisable.

A `fee` will often be calculated in the `flashLoan` function, which the `receiver` must be aware of for repayment. Passing the `fee` as a parameter instead of appended to `data` is simple and effective.

The `amount + fee` are pulled from the `receiver` to allow the `lender` to implement other functionality that depend on `token` balances, without having to lock it for the duration of a flash loan.

## Backwards Compatibility

No backwards compatibility issues identified.

## Implementation

### Flash Borrower Reference Implementation

```
pragma solidity ^0.7.5;

import "@openzeppelin/contracts/token/ERC20/IERC20.sol";

> import "@openzeppelin/contracts/math/SafeMath.sol";

Author: Fixed.

    }

    /**
     * @dev Loan `amount` tokens to `receiver`, and takes it back plus a `flashFee` after the callback.
     * @param receiver The contract receiving the tokens, needs to implement the `onFlashLoan(address user, uint256 amount, uint256 fee, bytes calldata)` interface.
     * @param token The loan currency.
     * @param amount The amount of tokens lent.
     * @param data A data parameter to be passed on to the `receiver` for any custom use.
     */
    function flashLoan(address receiver, address token, uint256 amount, bytes calldata data) external override {
        require(supportedTokens[token], "FlashLender: Unsupported currency");
        uint256 fee = _flashFee(token, amount);
        IERC20(token).transfer(receiver, amount);
        IERC3156FlashBorrower(receiver).onFlashLoan(msg.sender, token, amount, fee, data);

>         IERC20(token).transferFrom(receiver, address(this), amount.add(fee));

Reviewer 2: `transferFrom` doesn't necessarily throw on failure.  Sadly, some tokens don't return `true` on success either.  ERC20 spec says you **MUST** return `true` on success, so if you want to be spec compliant, wrap this call in a `require(...)`.  Same is true for the `.transfer(...)` call above.

Note: Will do a full review tomorrow, just wanted to call this one out early so you can get a head start on fixing it.

Reviewer 2: Alternatively, check balance before and after the call to make sure that the transfers went through successfully.  Downside is that it is another bunch of gas to do that.

Author: Fixed.

    }

    /**
     * @dev The fee to be charged for a given loan. Internal function with no checks.
     * @param token The loan currency.
     * @param amount The amount of tokens lent.
     * @return The amount of `token` to be charged for the loan, on top of the returned principal.
     */
    function _flashFee(address token, uint256 amount) internal view returns (uint256) {
        return fee == type(uint256).max ? 0 : amount.div(fee);
    }

    /**
     * @dev The amount of currency available to be lended.
     * @param token The loan currency.
     * @return The amount of `token` that can be borrowed.
     */
    function maxFlashAmount(address token) external view override returns (uint256) {
        return supportedTokens[token] ? IERC20(token).balanceOf(address(this)) : 0;
    }
}
```

## Security Considerations

### Verification of callback arguments

The arguments of `onFlashLoan` are expected to reflect the conditions of the flash loan, but cannot be trusted unconditionally. They can be divided in two groups, that require different checks before they can be trusted to be genuine.

> 0. No arguments can be assumed to be genuine without some kind of verification. `sender`, `token` and `value` refer to a past transaction that might not have happened if the caller of `onFlashLoan` decides to lie. `fee` might be false or calculated incorrectly. `calldata` is not expected to have been verified or manipulated by the caller.

Author: ```suggestion
0. No arguments can be assumed to be genuine without some kind of verification. `sender`, `token` and `value` refer to a past transaction that might not have happened if the caller of `onFlashLoan` decides to lie. `fee` might be false or calculated incorrectly. `data` might have been manipulated by the caller.
```

    function _flashFee(address token, uint256 amount) internal view returns (uint256) {

>         return fee == type(uint256).max ? 0 : amount.div(fee);

Author: Fixed into being defined as bps (multiples of 0.0001 %). I was being too smart.
